File: internal/sdk/context.go
package sdk

import (
	"context"
	"fmt"
	"net/url"
)

// User represents a ServiceNow user.
type User struct {
	SysID    string `json:"sys_id"`
	UserName string `json:"user_name"`
	Name     string `json:"name"`
	Email    string `json:"email"`
}

// GetCurrentUser retrieves the current user's info.
func (c *Client) GetCurrentUser(ctx context.Context) (*User, error) {
	params := url.Values{}
	params.Set("sysparm_limit", "1")
	params.Set("sysparm_fields", "sys_id,user_name,name,email")
	params.Set("sysparm_query", "sys_id=javascript:gs.getUserID()")

	records, err := c.List(ctx, "sys_user", params)
	if err != nil {
		return nil, fmt.Errorf("failed to query current user: %w", err)
	}

	if len(records) == 0 {
		return nil, fmt.Errorf("could not determine current user from session")
	}

	return userFromRecord(records[0]), nil
}

// userFromRecord converts a record map to a User struct.
func userFromRecord(record map[string]any) *User {
	return &User{
		SysID:    getString(record, "sys_id"),
		UserName: getString(record, "user_name"),
		Name:     getString(record, "name"),
		Email:    getString(record, "email"),
	}
}

// UpdateSet represents a ServiceNow update set.
type UpdateSet struct {
	SysID string `json:"sys_id"`
	Name  string `json:"name"`
}

// GetCurrentUpdateSet retrieves the current update set for the user from sys_user_preference.
func (c *Client) GetCurrentUpdateSet(ctx context.Context, userID string) (*UpdateSet, error) {
	// First, get the update set sys_id from user preferences
	query := url.Values{}
	query.Set("sysparm_limit", "1")
	query.Set("sysparm_fields", "value")
	query.Set("sysparm_query", "user="+userID+"^name=sys_update_set")

	prefRecords, err := c.List(ctx, "sys_user_preference", query)
	if err != nil {
		return nil, fmt.Errorf("failed to query user preference: %w", err)
	}

	// No preference set - return default
	if len(prefRecords) == 0 {
		return &UpdateSet{Name: "-"}, nil
	}

	// Get the update set sys_id from the preference value
	updateSetSysID := getString(prefRecords[0], "value")
	if updateSetSysID == "" {
		return &UpdateSet{Name: "-"}, nil
	}

	// Now fetch the update set name
	usQuery := url.Values{}
	usQuery.Set("sysparm_limit", "1")
	usQuery.Set("sysparm_fields", "sys_id,name")
	usQuery.Set("sysparm_query", "sys_id="+updateSetSysID)

	usRecords, err := c.List(ctx, "sys_update_set", usQuery)
	if err != nil {
		return nil, fmt.Errorf("failed to query update set details: %w", err)
	}

	if len(usRecords) == 0 {
		// Preference exists but update set was deleted - return default
		return &UpdateSet{Name: "-"}, nil
	}

	return &UpdateSet{
		SysID: getString(usRecords[0], "sys_id"),
		Name:  getString(usRecords[0], "name"),
	}, nil
}

// Application represents a ServiceNow application scope.
type Application struct {
	SysID string `json:"sys_id"`
	Scope string `json:"scope"`
	Name  string `json:"name"`
}

// GetCurrentApplication retrieves the current application scope for the user from sys_user_preference.
func (c *Client) GetCurrentApplication(ctx context.Context, userID string) (*Application, error) {
	// First, get the application sys_id from user preferences
	query := url.Values{}
	query.Set("sysparm_limit", "1")
	query.Set("sysparm_fields", "value")
	query.Set("sysparm_query", "user="+userID+"^name=apps.current_app")

	prefRecords, err := c.List(ctx, "sys_user_preference", query)
	if err != nil {
		return nil, fmt.Errorf("failed to query user preference: %w", err)
	}

	// No preference set - return global
	if len(prefRecords) == 0 {
		return &Application{Scope: "global"}, nil
	}

	// Get the application sys_id from the preference value
	appSysID := getString(prefRecords[0], "value")
	if appSysID == "" || appSysID == "global" {
		return &Application{Scope: "global"}, nil
	}

	// Now fetch the application scope details
	appQuery := url.Values{}
	appQuery.Set("sysparm_limit", "1")
	appQuery.Set("sysparm_fields", "sys_id,scope,name")
	appQuery.Set("sysparm_query", "sys_id="+appSysID)

	// Try sys_scope first, then sys_app
	appRecords, err := c.List(ctx, "sys_scope", appQuery)
	if err != nil || len(appRecords) == 0 {
		appRecords, err = c.List(ctx, "sys_app", appQuery)
		if err != nil {
			return nil, fmt.Errorf("failed to query application details: %w", err)
		}
	}

	if len(appRecords) == 0 {
		// Preference exists but application was deleted - return global
		return &Application{Scope: "global"}, nil
	}

	return &Application{
		SysID: getString(appRecords[0], "sys_id"),
		Scope: getString(appRecords[0], "scope"),
		Name:  getString(appRecords[0], "name"),
	}, nil
}

File: src/context.js
// ServiceNow runtime context: current user, scope, update set

import { getStringField } from './helpers.js';

export async function getCurrentUser(sdk) {
  const params = new URLSearchParams();
  params.set('sysparm_query', 'user_name=javascript:gs.getUserName()');
  params.set('sysparm_limit', '1');
  params.set('sysparm_display_value', 'all');
  params.set('sysparm_fields', 'sys_id,user_name,name');
  const records = await sdk.list('sys_user', params);
  if (records.length === 0) return null;
  const r = records[0];
  return {
    sys_id: r.sys_id?.value || r.sys_id,
    user_name: r.user_name?.display_value || r.user_name,
    name: r.name?.display_value || r.name,
  };
}

export async function getCurrentApplication(sdk, userSysID) {
  const params = new URLSearchParams();
  params.set('sysparm_query', `user=${userSysID}^name=apps.current_app`);
  params.set('sysparm_limit', '1');
  params.set('sysparm_fields', 'value');
  const records = await sdk.list('sys_user_preference', params);
  if (records.length === 0) return { scope: 'global', appSysId: '' };
  const val = records[0].value?.value || records[0].value;
  if (!val) return { scope: 'global', appSysId: '' };

  // Try to resolve sys_id to scope name
  try {
    const app = await sdk.get('sys_scope', val);
    if (app) return { scope: app.scope || 'global', appSysId: val };
  } catch {
    // fallback
  }
  try {
    const app2 = await sdk.get('sys_app', val);
    if (app2) return { scope: app2.scope || 'global', appSysId: val };
  } catch {
    // fallback
  }
  return { scope: val, appSysId: val };
}

export async function getCurrentUpdateSet(sdk, userSysID) {
  const params = new URLSearchParams();
  params.set('sysparm_query', `user=${userSysID}^name=sys_update_set`);
  params.set('sysparm_limit', '1');
  params.set('sysparm_fields', 'value');
  const records = await sdk.list('sys_user_preference', params);
  if (records.length === 0) return null;
  const val = records[0].value?.value || records[0].value;
  if (!val || val === '-') return { name: 'Default', sys_id: '' };

  try {
    const us = await sdk.get('sys_update_set', val);
    if (us) return { name: us.name || val, sys_id: val };
  } catch {
    // fallback
  }
  return { name: val, sys_id: val };
}

/**
 * Find-or-create a sys_user_preference and set its value. Shared by the
 * scope/update-set "set" commands and create auto-set (was copy-pasted 3x).
 */
async function setUserPreference(sdk, userSysID, name, value) {
  const params = new URLSearchParams();
  params.set('sysparm_query', `user=${userSysID}^name=${name}`);
  params.set('sysparm_limit', '1');
  params.set('sysparm_fields', 'sys_id');
  const prefs = await sdk.list('sys_user_preference', params);
  if (prefs.length > 0) {
    await sdk.update('sys_user_preference', getStringField(prefs[0], 'sys_id'), { value });
  } else {
    await sdk.create('sys_user_preference', {
      user: userSysID,
      name,
      value,
      type: 'string',
    });
  }
}

/** Resolve the current user's sys_id (throws if it can't be determined). */
export async function requireCurrentUserSysId(sdk) {
  const user = await sdk.list('sys_user', new URLSearchParams({
    sysparm_query: 'user_name=javascript:gs.getUserName()',
    sysparm_limit: '1',
    sysparm_fields: 'sys_id',
  }));
  if (user.length === 0) {
    throw new Error('Could not determine current user');
  }
  return getStringField(user[0], 'sys_id');
}

/** Set the current application scope via the apps.current_app preference. */
export async function setCurrentApplication(sdk, userSysID, scopeSysID) {
  await setUserPreference(sdk, userSysID, 'apps.current_app', scopeSysID);
}

/** Set the current update set via the sys_update_set preference. */
export async function setCurrentUpdateSet(sdk, userSysID, updateSetSysID) {
  await setUserPreference(sdk, userSysID, 'sys_update_set', updateSetSysID);
}
